operations in applyVariables

larkob opened the pull request:
This PR allows to perform COMPUTE operations in the `applyVariables` section of clickRoutines. 

Very basic example:
```
Operation using COMPUTE:
    {
      "func": "COMPUTE",
      "operation": "+",
      "operand1": 10,
      "operand2": 20,
      "variable": "COMPUTE"
    },
    {
      "func": "SET",
      "property": "y",
      "applyVariables": [
        {
          "parameter": "value",
          "variable": "COMPUTE"
        }
      ]
    }

Same with operation in applyVariables:
    {
      "func": "SET",
      "property": "y",
      "applyVariables": [
        {
          "operation": "+",
          "operand1": 10,
          "operand2": 20,
          "variable": "COMPUTE"
        },
        {
          "parameter": "value",
          "variable": "COMPUTE"
        }
      ]
    }
```

Another new feature is the use of variables for any of the three operands. You can add the prefix "var_" to the operand you want to mark as a variable:
```
    {
      "func": "GET",
      "property": "y",
      "variable": "myVariable"
    },
    {
      "func": "SET",
      "property": "y",
      "applyVariables": [
        {
          "operation": "+",
          "var_operand1": "myVariable",
          "operand2": 20,
          "variable": "COMPUTE"
        },
        {
          "parameter": "value",
          "variable": "COMPUTE"
        }
      ]
    }
```

Additionally, applyVariables now makes use of a special variable called "TEMP". After each applyVariables action is evaluated, TEMP is set to the resulting value, so it can be used as the input for the subsequent applyVariables action. At the end of the applyVariables actions, TEMP is deleted. With the TEMP variable and variable operands, you can for instance perform multiple operations like determining a random value between 1 and 100 before applying the value to a property:

```
   {
      "func": "SET",
      "property": "y",
      "relation": "=",
      "collection": "thisButton",
      "applyVariables": [
        {
          "operation": "random"
        },
        {
          "operation": "*",
          "var_operand1": "TEMP",
          "operand2": 100
        },
        {
          "operation": "round",
          "var_operand1": "TEMP"
        },
        {
          "operation": "+",
          "var_operand1": "TEMP",
          "operand2": 1
        },
        {
          "variable": "TEMP",
          "parameter": "value"
        }
      ]
    }
```

ArnoldSmith86 commented:
PR-SERVER-BOT: You can play around with it here: http://212.47.248.129:3293/pr-test (or any other room on that server)

96LawDawg commented:
So this is cool, as long as the original capability stays also.  This makes the code tighter and more flexible, but it increases the complexity.  More variables within applyVariables?  Definitely only experienced game designers are going to be able to understand and use this.

I tested it out using the COMPUTE tutorial in [http://212.47.248.129:3293/LawDawg](http://212.47.248.129:3293/LawDawg).  Only the actions in the white box on the left work.  And they only do multiplication because this tutorial was set up to make the COMPUTE "operation" itself a variable.  So that much of the code has been validated through my testing.

There may not be a use case for it beyond this tutorial, but would it be easily possible to turn "operation" into "var_operation" like you did "operand1" into "var_operand1"?  I could see some games needing to programmatically determine which operand to apply using variables (I add something this time and I subtract it next time).  This COMPUTE shortcut would only work if "operation" could only be a variable as well.

rogerl50 commented:
A couple of questions:
1. Does the ability to put the `COMPUTE` inside another function significantly shorten the code or improve the readability? Judging solely from the first example, it seems to me that the original formulation is just as clear, and almost as short, as the second.
2. Does the ability to prefix an operand name with `var_`, and the availability of the `TEMP` variable, apply within normal `COMPUTE`s as well (that is, both of the examples show the `COMPUTE` nested inside a `SET`; I want to make sure that I can do this with a standalone sequence of `COMPUTE`s).

larkob commented:
@96LawDawg I've fixed the bug in COMPUTE, to the center box in your test room works now. I've also added `var_operation`, although I'm not sure how useful this is. Subtraction is the same as adding a negative value, so there's no need for a computed operation.

larkob commented:
@rogerl50 
> A couple of questions:
> 
>     1. Does the ability to put the `COMPUTE` inside another function significantly shorten the code or improve the readability? Judging solely from the first example, it seems to me that the original formulation is just as clear, and almost as short, as the second.

I fully agree, but that's exactly why I used it as a first example, so everyone can see how the functionality works. If you try to replicate the third example using COMPUTE you'll see how much the readability is improved over multiple consecutive COMPUTE operations.
> 
>     2. Does the ability to prefix an operand name with `var_`, and the availability of the `TEMP` variable, apply within normal `COMPUTE`s as well (that is, both of the examples show the `COMPUTE` nested inside a `SET`; I want to make sure that I can do this with a standalone sequence of `COMPUTE`s).

You can already use applyVariables with COMPUTE operations, so offering sth. like a "var_operand1" in COMPUTE would not only create redundancy, but also lose the benefit of having a TEMP variable for the continuous calculations.

96LawDawg commented:
> I've also added `var_operation`, although I'm not sure how useful this is. Subtraction is the same as adding a negative value, so there's no need for a computed operation.

Thank you.  Either this didn't work (it probably did) or I can't test it because the test room needs to be reset after adding more commits.  Does anyone know if that is how this works?  I think I've seen this before: once the test server is created, no further changes to the code are reflected in the test room.  If that is correct, how do you trigger a refresh in the room so that latest commits get added?

Also, in response to @larkob, it isn't necessarily about doing a computed operation.  It is about not knowing in advance which computation you need to perform (add, subtract, divide, multiply, exponent, etc.).

larkob commented:
@96LawDawg I simply refresh the pr-test rooms with F5 after making a code change.

96LawDawg commented:
> @96LawDawg I simply refresh the pr-test rooms with F5 after making a code change. But there was actually a bug I needed to fix in order to make `var_operation` work.

Good to know.  And thanks.  It is now working in my COMPUTE tutorial test room.

96LawDawg reviewed (approved):
Approved through test.

rogerl50 commented:
Wouldn't it be useful to allow this construct in `COMPUTE` as well? Something like
```
  {
      "func": "COMPUTE",
      "operationList": [
        {
          "operation": "random"
        },
        {
          "operation": "*",
          "var_operand1": "TEMP",
          "operand2": 100
        },
        {
          "operation": "round",
          "var_operand1": "TEMP"
        },
        {
          "operation": "+",
          "var_operand1": "TEMP",
          "operand2": 1,
          "variable": "value"
        }
      ]
    }
```
Or are you proposing that there is a different way to do this using these capabilities? Perhaps I'm just not quite clear on exactly what the new functionality looks like and where it is applicable.

larkob commented:
I did consider it, but I concluded that there's no actual need to add this to COMPUTE. You could achieve the same with applyVariables now:
```
{
      "func": "COMPUTE",
      "operation": "=",
      "variable": "myResult",
      "applyVariables": [
        {
          "operation": "random"
        },
        {
          "operation": "*",
          "var_operand1": "TEMP",
          "operand2": 100
        },
        {
          "operation": "round",
          "var_operand1": "TEMP"
        },
        {
          "operation": "+",
          "var_operand1": "TEMP",
          "operand2": 1
        },
        {
          "variable": "TEMP",
          "parameter": "operand2"
        }
      ]
    }
```
There's a lot more power to the enhanced applyVariables than what I demonstrated with the examples above, but the wiki and some tutorials would probably be a better place than this PR to explain them.

I did consider allowing to use the "var_" prefix in other contexts than applyVariables, but without the additional benefits of the TEMP variable, it would only serve as a shorthand to the use of the regular applyVariables.

rogerl50 commented:
Well, I'm completely confused. But as long as it doesn't break existing code, I'll wait for the wiki writeup.

ArnoldSmith86 commented on client/js/widgets/button.js:
```diff
@@ -44,22 +44,186 @@ export class Button extends Widget {
+          Object.entries(v).forEach(([key, value]) => {
+            op[key[key.length-1]] = value;
+            if (['var_operand1', 'var_operand2', 'var_operand3'].indexOf(key) >= 0)
+              op[key[key.length-1]] = variables[value].index !== undefined ? variables[value][value.index] : variables[value];
```
This line looks very broken to me. You check if the given variable has an `index` and then use an `index` of the `value`. Is it even planned to use an index here? Seems too complicated.

larkob replied:
This wasn't added by me. I remember this was added a while back, but I cannot remember the exact use case.

larkob replied:
Oh wait, I think I see where I messed up.

larkob replied:
@ArnoldSmith86 With introducing compute operations in applyVariables, we actually no longer need the index parameter, as we could simply use getIndex as a prior compute step. I wonder if that's safe to remove the index parameter as it's never been documented and quite unlikely that any game uses it?

ArnoldSmith86 replied:
I'll check if it's being used anywhere on the server currently. Might be save.

rogerl50 replied:
> With introducing compute operations in applyVariables, we actually no longer need the index parameter, as we could simply use getIndex as a prior compute step. I wonder if that's safe to remove the index parameter as it's never been documented and quite unlikely that any game uses it?

I think that the original impetus for adding this was a request from me regarding a need I had in Dots. I worked around it and am happy with the result, so I at least never ended up using it.

ArnoldSmith86 commented on client/js/widgets/button.js:
```diff
@@ -44,22 +44,186 @@ export class Button extends Widget {
+            if (['var_operand1', 'var_operand2', 'var_operand3'].indexOf(key) >= 0)
+              op[key[key.length-1]] = variables[value].index !== undefined ? variables[value][value.index] : variables[value];
+          });
+          variables['TEMP'] = this.compute(v.operation, v.variable, op[1], op[2], op[3],problems);
```
Maybe call it LAST or PREVIOUS or something like that?

It's probably breaking if the variables for the operands didn't exist.

larkob replied:
Yeah, I had considered a bunch of different names, like LAST, PREVIOUS, CURRENT, INTERIM, RESULT and ended up with TEMP because it better conveys the idea that this variable is meant to be used only transitory during the applyVariables scope. I'm ok with changing the name to whatever people feel most adequate.

ArnoldSmith86 reviewed (changes requested):
Minor problems.

larkob commented on client/js/widgets/button.js:
```diff
@@ -165,155 +329,9 @@ export class Button extends Widget {
-
       if(a.func == 'COMPUTE') {
         setDefaults(a, { operation: '+', operand1: 1, operand2: 1, operand3: 1, variable: 'COMPUTE' });
-        const toNum = s=>typeof s == 'string' && s.match(/^[-+]?[0-9]+(\.[0-9]+)?$/) ? +s : s;
```
It's simply moved to line 82 (the compute method) so it's not limited to the COMPUTE operation only.

ArnoldSmith86 replied:
Oh, sorry. I explicitly checked if you put it into the function. Guess I checked the removed one.

larkob commented:
This is superseded by PR #307
